src/deepsparse/loggers/function_logger.py

```diff
@@ -0,0 +1,111 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Implementation of the Function Logger
+"""
+
+from typing import Any, Dict, List
+
+from deepsparse.loggers import BaseLogger, MetricCategories
+from deepsparse.loggers.metric_functions import apply_function
+
+
+__all__ = ["FunctionLogger"]
+
+
+class FunctionLogger(BaseLogger):
+    """
+    DeepSparse logger that applies functions to raw log values
+    according to the specified config file
+
+    :param loggers: A DeepSparse Logger object
+    :param config: A configuration dictionary that specifies the mapping between
+        the target name and the functions to be applied to raw log values
+        e.g.
+        ```
+        {"pipeline_inputs":
+            [{"function": "identity",
+              "frequency": 3},
+
+            [{"function": "identity",
+              "frequency": 5}],
+
+         "pipeline_outputs":
+            [{"function": "identity",
+            "frequency": 4}]
+        }
+    """
+
+    def __init__(
+        self,
+        logger: BaseLogger,
+        config: Dict[str, List[Dict[str, Any]]],
+    ):
+
+        self.logger = logger
+        self.config = self._add_frequency_counter(config)
+
+    def log(self, identifier: str, value: Any, category: MetricCategories):
+        """
+        :param identifier: The identifier of the log
+             By default should consist of at least one string (name of the pipeline):
+                e.g. identifier = "pipeline_name"
+            If identifier is to be more complex, optional strings are to be concatenated
+            using dot as a separator:
+                e.g. identifier = "pipeline_name.some_argument_1.some_argument_2"
+        :param value: The data structure that the logger is logging
+        :param category: The metric category that the log belongs to
+        """
+        if category == MetricCategories.DATA:
+            pipeline_name, *target_name = identifier.split(".")
```

Comment:
for now, let's assume that the config includes the fully-qualified name and not strip out the pipeline name from identifier - this builds too much into the server config scenario and doesn't allow for one logger to handle multiple pipelines well.

Reply:
So maybe let's enforce this "decomposition" explicitly using the args of the `abstractmethod` in `BaseLogger(ABC)`:

```python
@abstractmethod
    def log(self, pipeline_name: str, target: str, value: Any, category: Optional[str] = None):
```
a then `pipeline` would be something like `image_classification` or `my_little_pipeline`
while `target` could be `engine_outputs` as well as `super.extra.dalle.latent.vectors`?

Reply:
that's would tie the use of log to pipelines a bit too much for where we are at in the implementation phase. let's see where we end up with configs, especially with multiple models and go from there.

Reply:
but isn't it what Loggers are in the end? Attributes of pipelines?